pythonforandroid/toolchain.py

```diff
@@ -137,38 +149,67 @@ def shprint(command, *args, **kwargs):
     kwargs["_out_bufsize"] = 1
     kwargs["_err_to_out"] = True
     kwargs["_bg"] = True
+    is_critical = kwargs.pop('_critical', False)
+    tail_n = kwargs.pop('_tail', 0)
+    filter_in = kwargs.pop('_filter', None)
+    filter_out = kwargs.pop('_filterout', None)
     if len(logger.handlers) > 1:
         logger.removeHandler(logger.handlers[1])
+    try:
+        columns = max(25, int(os.popen('stty size', 'r').read().split()[1]))
+    except:
+        columns = 100
     command_path = str(command).split('/')
     command_string = command_path[-1]
     string = ' '.join(['running', command_string] + list(args))
 
     # If logging is not in DEBUG mode, trim the command if necessary
     if logger.level > logging.DEBUG:
-        short_string = string
-        if len(string) > 100:
-            short_string = string[:100] + '... (and {} more)'.format(len(string) - 100)
-        logger.info(short_string + Err_Style.RESET_ALL)
+        logger.info('{}{}'.format(shorten_string(string, columns - 12), Err_Style.RESET_ALL))
     else:
-        logger.debug(string + Err_Style.RESET_ALL)
+        logger.debug('{}{}'.format(string, Err_Style.RESET_ALL))
 
-    output = command(*args, **kwargs)
     need_closing_newline = False
-    for line in output:
-        if logger.level > logging.DEBUG:
-            string = ''.join([Err_Style.RESET_ALL, '\r', ' '*11, 'working ... ',
-                              line[:100].replace('\n', '').rstrip(), ' ...'])
-            if len(string) < 20:
-                continue
-            if len(string) < 120:
-                string = string + ' '*(120 - len(string))
-            sys.stdout.write(string)
-            sys.stdout.flush()
-            need_closing_newline = True
+    try:
+        msg_hdr = '           working: '
+        msg_width = columns - len(msg_hdr) - 1
+        output = command(*args, **kwargs)
+        for line in output:
+            if logger.level > logging.DEBUG:
+                msg = line.replace('\n', ' ').replace('\t', ' ').replace('\b', ' ').rstrip()
+                if msg:
+#                    if len(msg) > msg_width: msg = msg[:(msg_width - 3)] + '...'
+                    sys.stdout.write('{}\r{}{:<{width}}'.format(Err_Style.RESET_ALL, msg_hdr, shorten_string(msg, msg_width), width=msg_width))
+                    sys.stdout.flush()
+                    need_closing_newline = True
+            else:
+                logger.debug(''.join(['\t', line.rstrip()]))
+        if need_closing_newline: sys.stdout.write('{}\r{:>{width}}\r'.format(Style.RESET_ALL, ' ', width=(columns - 1)))
+    except sh.ErrorReturnCode, err:
```

Comment:
Minor comment: this would be better as `except sh.ErrorReturnCode as err' - especially since I don't think python3 supports the comma syntax any more.